sql/rpl_info_file.h

```diff
@@ -0,0 +1,322 @@
+  {
+    /**
+      my_b_printf() uses a buffer too,
+      so we might as well save on format parsing and buffer resizing
+    */
+    char buf[BUF_SIZE<I>];
+    std::to_chars_result result= std::to_chars(buf, &buf[sizeof(buf)], value);
+    DBUG_ASSERT(result.ec == ERRC_OK);
+    my_b_write(file, reinterpret_cast<const uchar *>(buf), result.ptr - buf);
+  }
+};
+
+
+/**
+  This common superclass of @ref Master_info_file and
+  @ref Relay_log_info_file provides them common code for saving
+  and loading fields in their MySQL line-based sections.
+  As only the @ref Master_info_file has a MariaDB `key=value`
+  section with a mix of explicit and `DEFAULT`-able fields,
+  code for those are in @ref Master_info_file instead.
+
+  Each field is an instance of an implementation
+  of the @ref Info_file::Persistent interface.
+  C++ templates enables code reuse for those implementation structs, but
+  templates are not suitable for the conventional header/implementation split.
+  Thus, this and derived files are header-only units (methods are `inline`).
+  Other files may include these files directly,
+  though headers should include this set under their `#include` guards.
+  [C++20 modules](https://en.cppreference.com/w/cpp/language/modules.html)
+  can supercede headers and their `#include` guards.
+*/
+struct Info_file
+{
+  IO_CACHE file;
+
+
+  /// Persistence interface for an unspecified item
+  struct Persistent
+  {
+    virtual ~Persistent()= default;
+    // for save_to_file()
+    virtual bool is_default() { return false; }
+    /// @return `true` if the item is mandatory and couldn't provide a default
+    virtual bool set_default() { return true; }
+    /** set the value by reading a line from the IO and consume the `\n`
+      @return `false` if the line has parsed successfully or `true` if error
+      @post is_default() is `false`
+    */
+    virtual bool load_from(IO_CACHE *file)= 0;
+    /** write the *effective* value to the IO **without** a `\n`
+      (The caller will separately determine how
+      to represent using the default value.)
+    */
+    virtual void save_to(IO_CACHE *file)= 0;
+  };
+
+  /** Integer Field
+    @tparam I signed or unsigned integer type
+    @see Master_info_file::Optional_int_field
+      version with `DEFAULT` (not a subclass)
+  */
+  template<typename I> struct Int_field: Persistent
+  {
+    I value;
+    operator I() { return value; }
+    auto &operator=(I value)
+    {
+      this->value= value;
+      return *this;
+    }
+    virtual bool load_from(IO_CACHE *file) override
+    { return Int_IO_CACHE::from_chars(file, value); }
+    virtual void save_to(IO_CACHE *file) override
+    { return Int_IO_CACHE::to_chars(file, value); }
+  };
+
+  /// Null-Terminated String (usually file name) Field
+  template<size_t size= FN_REFLEN> struct String_field: Persistent
+  {
+    char buf[size];
+    virtual operator const char *() { return buf; }
```

Comment:
Please add a comment to explain why this operator override is needed (i.e how it is used by other sub-classes)